abstract_domains/numerical/interval.py

```diff
@@ -0,0 +1,242 @@
+            if expr.operator == BinaryArithmeticOperation.Operator.Add:
+                return l.add(r)
+            elif expr.operator == BinaryArithmeticOperation.Operator.Sub:
+                return l.sub(r)
+            elif expr.operator == BinaryArithmeticOperation.Operator.Mult:
+                return l.mult(r)
+            else:
+                raise ValueError(f"Binary Operator '{str(expr.operator)}' is not supported!")
+
+        def visit_UnaryArithmeticOperation(self, expr: UnaryArithmeticOperation):
+            r = self.visit(expr.expression)
+            if expr.operator == UnaryArithmeticOperation.Operator.Add:
+                return r
+            elif expr.operator == UnaryArithmeticOperation.Operator.Sub:
+                return r.negate()
+            else:
+                raise ValueError(f"Unary Operator {expr.operator} is not supported!")
+
+        def visit_Literal(self, expr: Literal):
+            if expr.typ == int:
+                c = int(expr.val)
+                return IntervalLattice(c, c)
+            else:
+                raise ValueError(f"Literal type {expr.typ} is not supported!")
+
+    _visitor = Visitor()  # static class member shared between all instances
+
+
+class IntervalStore(Store, NumericalDomain):
+    def __init__(self, variables: List[VariableIdentifier]):
+        super().__init__(variables, {int: IntervalLattice})
+        self._visitor = IntervalStore.Visitor(self)
+
+    def forget(self, var: VariableIdentifier):
+        self.store[var].top()
+
+    def set_bounds(self, var: VariableIdentifier, lower: int, upper: int):
+        self.store[var].lower = lower
+        self.store[var].upper = upper
+
+    def get_bounds(self, var: VariableIdentifier):
+        return self.store[var].lower, self.store[var].upper
+
+    def set_interval(self, var: VariableIdentifier, interval: IntervalLattice):
+        self.store[var].lower = interval.lower
+        self.store[var].upper = interval.upper
+
+    def set_lb(self, var: VariableIdentifier, constant):
+        self.store[var].lower = constant
+
+    def set_ub(self, var: VariableIdentifier, constant):
+        self.store[var].upper = constant
+
+    def evaluate(self, expr: Expression):
+        interval = self._visitor.visit(expr)
+        return interval
+
+    class Visitor(IntervalLattice.Visitor):
+        """A visitor to abstractly evaluate an expression (with variables) in the interval domain."""
+
+        def __init__(self, interval_store):
+            self._store = interval_store  # keep a reference to container interval store
+
+        @property
+        def store(self):
+            return self._store
+
+        def visit_VariableIdentifier(self, expr: VariableIdentifier):
+            if expr.typ == int:
+                return self.store.store[expr]
+            else:
+                raise ValueError(f"Variable type {expr.typ} is not supported!")
+
+
+class IntervalDomain(IntervalStore, State):
```

Comment:
I do not understand why IntervalStore and IntervalDomain are separate. I would expect that IntervalDomain is a Store lattice that can be used as a State.